### PR TITLE
Move tcl keyspace test cases to Go

### LINKED ISSUE #841: Move TCL test unit/keyspace to Go case

### PR DESCRIPTION
**Pull request by git-hulk**
This closes #841 

I think it's a bit wordy to add the directory for only one file, so I used the `keyspace_test.go` instead of `keyspace/keyspace_test.go`

### CODE REVIEW

File: tests/gocase/unit/keyspace_test.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+package unit
+
+import (
+	"context"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go]: I suggest always using `require` to avoid unnecessary dirty states. We just fail fast.

File: tests/gocase/unit/keyspace_test.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	defer srv.Close()
+
+	ctx := context.Background()
+	rdb := srv.NewClient()
+	defer func() { require.NoError(t, rdb.Close()) }()
+
+	t.Run("DEL against a single item", func(t *testing.T) {
+		key := "x"
+		value := "foo"
+		assert.Nil(t, rdb.Set(ctx, key, value, 0).Err())
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go]:
```suggestion
		require.NoError(t, rdb.Set(ctx, key, value, 0).Err())
```

Comment (tisonkun) [reply tests/gocase/unit/keyspace_test.go]: ditto others

**Review by tisonkun — CHANGES_REQUESTED** [review]
Two global style comments. Will give a functionality review after they're addressed.

File: tests/gocase/unit/keyspace_test.go (hunk 1 above)
Comment (git-hulk) [reply tests/gocase/unit/keyspace_test.go]: Fixed, you're right that previous steps failed will affect the successor steps in most test cases, so it's better to use require instead of assert.

File: tests/gocase/unit/keyspace_test.go (hunk 2 above)
Comment (git-hulk) [reply tests/gocase/unit/keyspace_test.go]: all replace with `NoError`

File: tests/gocase/unit/keyspace_test.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+			return keys[i] < keys[j]
+		})
+		require.Equal(t, []string{"foo_a", "foo_b", "foo_c", "key_x", "key_y", "key_z"}, keys)
+	})
+
+	t.Run("DEL all keys", func(t *testing.T) {
+		vals := rdb.Keys(ctx, "*").Val()
+		require.Equal(t, int64(len(vals)), rdb.Del(ctx, vals...).Val())
+		require.NoError(t, rdb.Do(ctx, "dbsize", "scan").Err())
+		time.Sleep(100 * time.Millisecond)
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go:86]: Does `dbsize scan` asynchorously? Why we need to sleep here?

File: tests/gocase/unit/keyspace_test.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	})
+
+	t.Run("KEYS to get all keys", func(t *testing.T) {
+		keys := rdb.Keys(ctx, "*").Val()
+		sort.Slice(keys, func(i, j int) bool {
+			return keys[i] < keys[j]
+		})
+		require.Equal(t, []string{"foo_a", "foo_b", "foo_c", "key_x", "key_y", "key_z"}, keys)
+	})
+
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go:75]: Missing

```tcl
    test {DBSIZE} {
        r dbsize scan
        after 100
        r dbsize
    } {6}
```

Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go:75]: Missing

```tcl
    test {DBSIZE} {
        r dbsize scan
        after 100
        r dbsize
    } {6}
```

Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go:75]: Missing

```tcl
    test {DBSIZE} {
        r dbsize scan
        after 100
        r dbsize
    } {6}
```

File: tests/gocase/unit/keyspace_test.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+		time.Sleep(100 * time.Millisecond)
+		require.Equal(t, int64(0), rdb.Do(ctx, "dbsize").Val())
+	})
+
+	t.Run("EXISTS", func(t *testing.T) {
+		newKey := "newkey"
+		require.NoError(t, rdb.Set(ctx, newKey, "test", 0).Err())
+		require.Equal(t, int64(1), rdb.Exists(ctx, newKey).Val())
+		require.Equal(t, int64(1), rdb.Del(ctx, newKey).Val())
+		require.Equal(t, int64(0), rdb.Exists(ctx, newKey).Val())
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go]: If you use `require.EqualValues`, you don't need to cast between integers. But this is no a requirement. Just for your information and current style is OK for me, too.

File: tests/gocase/unit/keyspace_test.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+		r, err = c.ReadLine()
+		require.NoError(t, err)
+		require.Equal(t, "xyzk", r)
+		r, err = c.ReadLine()
+		require.NoError(t, err)
+		require.Equal(t, "+PONG", r)
+	})
+
+	t.Run("Non existing command", func(t *testing.T) {
+		require.Contains(t, rdb.Do(ctx, "foobaredcommand").Err().Error(), "ERR")
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go]:
```suggestion
		util.ErrorRegexp(t, rdb.Do(ctx, "foobaredcommand").Err(), "ERR.*")
```

File: tests/gocase/unit/keyspace_test.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+			randomKey := rdb.RandomKey(ctx).Val()
+			if randomKey == "foo" {
+				fooSeen = 1
+			} else if randomKey == "bar" {
+				barSeen = 1
+			}
+			if fooSeen == 1 || barSeen == 1 {
+				break
+			}
+			time.Sleep(10 * time.Millisecond)
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go]: Why do we need to sleep here?

File: tests/gocase/unit/keyspace_test.go [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+		}
+		require.Greater(t, fooSeen+barSeen, 0)
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go]: If you'd like to write as:

```go
			if fooSeen == 1 || barSeen == 1 {
				break
			}
```

I suggest:

```go
		for i := 0; i < 1000; i++ {
			randomKey := rdb.RandomKey(ctx).Val()
			switch randomKey {
			case "foo", "bar":
				return
			}
		}
		require.Fail(t, "RANDOMKEY never hits foo or bar")
```

File: tests/gocase/unit/keyspace_test.go [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+	})
+
+	t.Run("RANDOMKEY regression 1", func(t *testing.T) {
+		rdb.FlushDB(ctx)
+		require.NoError(t, rdb.Set(ctx, "x", 10, 0).Err())
+		require.Equal(t, int64(1), rdb.Del(ctx, "x").Val())
+		require.Equal(t, "", rdb.RandomKey(ctx).Val())
+	})
+
+	t.Run("KEYS * two times with long key, Github issue #1208", func(t *testing.T) {
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go]: This is for the Redis issue number?
```suggestion
	t.Run("KEYS * two times with long key - Redis Github issue #1208", func(t *testing.T) {
```

File: tests/gocase/unit/keyspace_test.go [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+		rdb.FlushDB(ctx)
+		require.NoError(t, rdb.Set(ctx, "dlskeriewrioeuwqoirueioqwrueoqwrueqw", "teset", 0).Err())
+		require.Equal(t, []string{"dlskeriewrioeuwqoirueioqwrueoqwrueqw"}, rdb.Keys(ctx, "*").Val())
```
Comment (tisonkun) [inline tests/gocase/unit/keyspace_test.go:159]:
```suggestion
		require.Equal(t, []string{"dlskeriewrioeuwqoirueioqwrueoqwrueqw"}, rdb.Keys(ctx, "*").Val())
		require.Equal(t, []string{"dlskeriewrioeuwqoirueioqwrueoqwrueqw"}, rdb.Keys(ctx, "*").Val())
```

I think the things to be tested here is "KEYS * two times"

**Review by tisonkun — CHANGES_REQUESTED** [review]
Thanks for your update. Comments inline.

File: tests/gocase/unit/keyspace_test.go (hunk 3 above)
Comment (git-hulk) [reply tests/gocase/unit/keyspace_test.go:86]: Yes, the DB scan is asynchronous, so need to wait a moment. Current behavior is consistent with tcl test cases.

File: tests/gocase/unit/keyspace_test.go (hunk 4 above)
Comment (git-hulk) [reply tests/gocase/unit/keyspace_test.go:75]: Good catch

File: tests/gocase/unit/keyspace_test.go (hunk 10 above)
Comment (git-hulk) [reply tests/gocase/unit/keyspace_test.go:159]: Yes, I saw this but I think this line should be duplicate.

Comment (tisonkun) [reply tests/gocase/unit/keyspace_test.go:159]: No. The case is that Redis once encounter an issue that if run `KEY *` twice, it will fail. I tend to keep these two lines or we just rename/remove the test case.

See also https://github.com/redis/redis/issues/1208.

File: tests/gocase/unit/keyspace_test.go (hunk 5 above)
Comment (git-hulk) [reply tests/gocase/unit/keyspace_test.go]: This's a good suggestion, have done.

File: tests/gocase/unit/keyspace_test.go (hunk 10 above)
Comment (git-hulk) [reply tests/gocase/unit/keyspace_test.go:159]: Got it, thanks. Just keep those two lines.

File: tests/gocase/unit/keyspace_test.go (hunk 7 above)
Comment (git-hulk) [reply tests/gocase/unit/keyspace_test.go]: We random by time seed, so it may always get the same seed if requests fast enough. But the case looks good until now, so let's remove the sleep.

Comment (git-hulk) [reply tests/gocase/unit/keyspace_test.go]: We random by time seed, so it may always get the same seed if requests fast enough. But the case looks good until now, so let's remove the sleep.

**Review by tisonkun — APPROVED** [review]
Looks good to me. Thank you!

**Comment by git-hulk** [comment]
Thanks for @tisonkun carefully review, will merge after the CI become green.
